### PR TITLE
Incorporate HCO's nodeselectors logic into KubeVirt

### PR DESCRIPTION
**Pull request by stu-gott**
**What this PR does / why we need it**:

HCO recently created logic to pass NodeSelectors, Affinity and Tolerations to each component. This PR incorporates that logic into KubeVirt

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
-->
```release-note
Allow specifying nodeSelectors, affinity and tolerations to control where KubeVirt components will run
```

### CODE REVIEW

File: staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go
Comment (xpivarc) [inline staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go:38]: Would you be so kind and use [marker](https://kubernetes.io/docs/reference/using-api/api-concepts/#merge-strategy) on this field. It will sillent kube-openapi.

Comment (stu-gott) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go:38]: Odd that k8s doesn't apply a marker to this field either?

@tiraboschi All HCO-based projects should do this the same way. Please advise.

File: staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go
Comment (stu-gott) [inline staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go:31]: @tiraboschi please note this is implemented as a pointer -- makes the code that needs to use this much cleaner. This differs from the HCO suggested defaults.

File: staging/src/kubevirt.io/client-go/api/v1/types.go
Comment (stu-gott) [inline staging/src/kubevirt.io/client-go/api/v1/types.go:1165]: @tiraboschi pointers for Infra and Workloads too...

**Comment by stu-gott** [comment]
/hold

Failing tests look very much like this PR caused them

File: pkg/virt-operator/install-strategy/create.go
Comment (davidvossel) [inline pkg/virt-operator/install-strategy/create.go:91]: just a reminder to remove the debug log lines.

File: staging/src/kubevirt.io/client-go/api/v1/types.go
Comment (davidvossel) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:1165]: is there a reason pointers are used here?

File: pkg/virt-operator/install-strategy/create.go
Comment (davidvossel) [inline pkg/virt-operator/install-strategy/create.go:331]: you won't need this `podSpecsMatch` function once #3612 is merged.

Ashley's pr ensures that the items in the install strategy are only updated if the "generationId" on the KubeVirt CR changes. This id only changes when the spec is updated. 

So once her PR goes in, we can be sure that your toleration,affinity,selector changes are applied as long as you inject them into the install strategy like you're doing here.

File: staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go
Comment (tiraboschi) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go:38]: `// +listType=set` ?

File: pkg/virt-operator/install-strategy/create.go
Comment (stu-gott) [reply pkg/virt-operator/install-strategy/create.go:91]: ouch lol. glad I left a note. thanks for the reminder

File: staging/src/kubevirt.io/client-go/api/v1/types.go
Comment (stu-gott) [reply staging/src/kubevirt.io/client-go/api/v1/types.go:1165]: a couple of reasons.

if these are static there's junk/nonsense lines in the kubevirt-cr `tolerations: {}` -- so it's cleaner and less confusing to admins.

If this is a pointer, the operator can short circuit the entire merge process at a glance. if this is static, there's no obvious queue that no values are/are not set and each one must be checked.

File: staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go
Comment (stu-gott) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go:38]: :rofl: I'll accept this as an authoritative answer despite the question mark...

File: staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go
Comment (maya-r) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go:31]: I imagine it's fine to make changes as long as we haven't made a release of CDI

Comment (nunnatsa) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go:31]: Done: https://github.com/kubevirt/hyperconverged-cluster-operator/pull/785

File: staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go
Comment (maya-r) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go:38]: I'm trying to sync CDI with the code in HCO.
I get slightly more errors with listType=set
````
Type: "FieldValueInvalid",
Message: "Invalid value: \"object\": must be a scalar or atomic type as item of a list with x-kubernetes-list-type=set",
Field: "spec.versions[1].schema.openAPIV3Schema.properties[spec].properties[workload].properties[tolerations].items.type",
````

Is there something about the HCO code that handles this issue?

File: pkg/virt-operator/install-strategy/create.go
Comment (stu-gott) [reply pkg/virt-operator/install-strategy/create.go]: Confirmed it's not needed after rebase. deleted. thanks!

File: staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go
Comment (xpivarc) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go:38]: I think it should be ` //+listType=map`

**Comment by stu-gott** [comment]
/cc @davidvossel 

I think this is ready for critical review

**Review by davidvossel — APPROVED** [review]
/lgtm
/approve

nice work!

File: staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go
Comment (rmohr) [inline staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go:15]: nit: This is actually also for KubeVirt the preferred way of shaping the placement over the customization-patch-way. I would therefore vote for a struct name like `ComponentPlacement` or the like.

File: api/api-rule-violations-known.list
Comment (rmohr) [inline api/api-rule-violations-known.list:150]: Are we still missing the right annotation on the map?

**Review by rmohr — CHANGES_REQUESTED** [review]
We said that we don't want to introduce more API rule validations, that should be fixed. Also I find the HCO name appearing in the structs and so on confusing. This should be the number-one entrypoint for anyone who wants to customize kubevirt in a way which will definitely work with updates and so on.

File: staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go
Comment (stu-gott) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go:15]: @tiraboschi was the name HyperConvergedConfig supposed to be literal for all projects? Is there an issue with calling this ComponentPlacement?

Comment (nunnatsa) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go:15]: > @tiraboschi was the name HyperConvergedConfig supposed to be literal for all projects? Is there an issue with calling this ComponentPlacement?

I think we need to use the new [SDK](https://github.com/kubevirt/controller-lifecycle-operator-sdk) for that. Currentlly we merged 3 diffrent implementation into HCO, and then we must also translate each one of them.

File: api/api-rule-violations-known.list
Comment (stu-gott) [reply api/api-rule-violations-known.list]: fixed

File: staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go
Comment (stu-gott) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go:15]: @nunnatsa you answer confused me somewhat. Are you suggesting that we need to re-architect this PR? Is this SDK required if we wanted to change the name? Would it still be something we needed if we followed suit and used HCO's naming convention?

Comment (nunnatsa) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go:15]: > Are you suggesting that we need to re-architect this PR?

I guess we don't have time for that. But I would do it in the future. 

> Is this SDK required if we wanted to change the name? 

No. This only for HCO to use the same go type for all the ipwrators,and to avoid translation when passong the data. 

Would it still be something we needed if we followed suit and used HCO's naming convention?

No. But it will be easier.

Comment (rmohr) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go:15]: The main issue is that this feature has nothing to do with HCO as such. It just appears that HCO wants to use it too (like numerous users which requested an easy way to influence the placement). We export the struct names with our kubevirt.io/client-go and it will also be in the api spec. This can be confusing.

Comment (stu-gott) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go:15]: @tiraboschi, @rmohr makes a very good point. Once this is merged/released it becomes pretty difficult to change. Is a more product-agnostic naming convention permissible here?

Comment (rmohr) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go:15]: We can change it later on. We don't guarantee struct-stability (only api stability). So I have no intent on blocking on that, but I would want to see it at least considered.

Comment (nunnatsa) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go:15]: So I moved HCO to use the sdk version and it was almost immediate change.

The naming there has nothing to do with HCO and I suggest agian to adopt it. the change is minor, because the field names and types are the same, because they also adopted HCO structure:

```golang
// NodePlacement describes node scheduling configuration.
// +k8s:openapi-gen=true
type NodePlacement struct {
	// nodeSelector is the node selector applied to the relevant kind of pods
	// It specifies a map of key-value pairs: for the pod to be eligible to run on a node,
	// the node must have each of the indicated key-value pairs as labels
	// (it can have additional labels as well).
	// See https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#nodeselector
	// +kubebuilder:validation:Optional
	// +optional
	NodeSelector map[string]string `json:"nodeSelector,omitempty"`

	// affinity enables pod affinity/anti-affinity placement expanding the types of constraints
	// that can be expressed with nodeSelector.
	// affinity is going to be applied to the relevant kind of pods in parallel with nodeSelector
	// See https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#affinity-and-anti-affinity
	// +kubebuilder:validation:Optional
	// +optional
	Affinity *corev1.Affinity `json:"affinity,omitempty"`

	// tolerations is a list of tolerations applied to the relevant kind of pods
	// See https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/ for more info.
	// These are additional tolerations other than default ones.
	// +kubebuilder:validation:Optional
	// +optional
	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
}
```
(source: https://github.com/kubevirt/controller-lifecycle-operator-sdk/blob/master/pkg/sdk/api/types.go#L49)

Comment (nunnatsa) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go:15]: The HCO PR: https://github.com/kubevirt/hyperconverged-cluster-operator/pull/801

Comment (rmohr) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go:15]: `NodePlacement` sounds better than `HyperConvergedConfig`. :+1:

Comment (nunnatsa) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go:15]: @rmohr - it's more than just naming. I'm talking about the same go struct.

**Comment by stu-gott** [comment]
Changed all references to HyperConvergedConfig to NodePlacement. rebased.

Unit tests failed for me locally. Perhaps related to the rebase as the failures are unrelated to this PR. I assume master is not broken?

File: staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go
Comment (tiraboschi) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go]: As usual, we debated a lot about namings also when we introduced the APIs on HCO side.
See for instance: https://github.com/kubevirt/hyperconverged-cluster-operator/pull/718#issuecomment-672973258

Initially it was exactly called placement also there, but then we decided to make it more generic and call it HyperConvergedConfig to let us inject more values to be propagated to components operator in the future.
So we decided to nest placement under that, then we decided to remove it because NodeSelector/Affinity/Tolerations was declarative enough.

Comment (rmohr) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go]: The discussion here is of course independent of the naming in HCO itself. Having Hyperconverged here in the struct name simply does not make sense imo. Simone, do I miss something?

Comment (tiraboschi) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go]: We named it HyperConvergedConfig on HCO CR to make more evident that it's content is going to be propagated to other components (so hyperconverged) and it's not HCO specific.
HyperConvergedConfig is just a container of various configuration options, NodePlacement is just the first one.

Comment (rmohr) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go]: Do you suggest to not name it NodePlacement? Do we agree that it does not make sense to introduce a struct with the name HyperConverged in the kubevirt operator? It just deploys kubevirt, there is nothing "hyperconverged" here.
I think I miss what you want to say.

Comment (nunnatsa) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go]: We want to use HCO as a proxy for configurations that relevant to all (or most of) the other operators. Right now it's node placement, but this name is limiting onlyu for node placement use cases.

Comment (rmohr) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go]: I can still not follow. All I care about is that it is not called `HyperConvergedSomething`, since it has nothing to do with hyperconverged in kubevirt. If you feel more comfortable with another name, just propose it.

Comment (rmohr) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go]: We can also again rename the struct later if we like, but it is obvious that `HyperConverged` is confusing and wrong in this context, therefore I want to avoid it.

Comment (nunnatsa) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go]: Not using "HyperConverged" is understanable. But since it was wrapped in other places as "NodePlacement" it's to specific for what it should do in HCO, so we nested it:

```golang
type HyperConvergedConfig struct {
	// NodePlacement describes node scheduling configuration.
	// +optional
	NodePlacement *sdkapi.NodePlacement `json:"nodePlacement,omitempty"`
}
```

Comment (rmohr) [reply staging/src/kubevirt.io/client-go/api/v1/hyperconvergedconfig.go]: We can nest it here too, sounds good. Something like `ComponentConfig` could be good enough.

**Comment by stu-gott** [comment]
Rebased and fixed merge conflicts. No other changes.

**Comment by stu-gott** [comment]
Ack! Sorry @davidvossel I didn't realize you were code reviewing the previous iteration. The code has been updated to wrap NodePlacement inside ComponentConfig, per @rmohr's request.

A fairly extensive change in terms of the code touched, but in terms of impact to logic or algorithm--not that much.

**Comment by stu-gott** [comment]
Thanks for catching that error @xpivarc

**Comment by stu-gott** [comment]
Thanks @rmohr!

**Comment by rmohr** [comment]
@stu-gott is this missing a user-guide entry?

**Pull request merged**
